[DOC-7519] Publish A-101963 (#16872)
cockroachdb/docs: advisories/a101963.md

---
title: Technical Advisory 101963
advisory: A-101963
summary: In specific circumstances, a RESTORE operation may restore incorrect data from a backup that contains incorrect metadata related to revision history.
toc: true
affected_versions: v22.2.0 to v22.2.8 and v22.1.0 to 22.1.19
advisory_date: 2023-05-09
docs_area: releases
---

Publication date: {{ page.advisory_date | date: "%B %e, %Y" }}

## Description

In the following circumstances, backups that include [revision history](../{{site.versions["stable"]}}/take-backups-with-revision-history-and-restore-from-a-point-in-time.html) could record incorrect metadata, and a subsequent `RESTORE` operation could write incorrect data.

> - In a cluster running an impacted v22.2 or v22.1 version, a backup of an `UPDATE`-heavy workload captures more than `kv.bulk_sst.max_allowed_overage` of revision history for a given row (64 MB by default). A subsequent `RESTORE` operation could erroneously omit or include a row, or could restore an incorrect version of it.

Author: I think I've got it now -- for some reason I found this really confusing. Thanks for your patience.

> - In a cluster running an impacted v22.2 or v22.1 version, a backup of an `UPDATE`-heavy workload captures more than `kv.bulk_sst.max_allowed_overage` of revision history for a given row (64 MB by default). A subsequent `RESTORE` operation could erroneously omit or include a row, or could restore an incorrect version of it.

Reviewer: LGTM

---
title: Technical Advisory 101963
advisory: A-101963
summary: In specific circumstances, a RESTORE operation may restore incorrect data from a backup that contains incorrect metadata related to revision history.
toc: true
affected_versions: v22.2.0 to v22.2.8 and v22.1.0 to 22.1.19
advisory_date: 2023-05-09
docs_area: releases
---

Publication date: {{ page.advisory_date | date: "%B %e, %Y" }}

## Description

> In the following two circumstances, backups that include [revision history](../stable/take-backups-with-revision-history-and-restore-from-a-point-in-time.html) may record incorrect metadata. The presence of this incorrect metadata could cause a subsequent `RESTORE` operation to write incorrect data:

Reviewer: ```suggestion
In the following two circumstances on CockroachDB v22.2.0 to v22.2.8 and v22.1.0 to 22.1.19, backups that include [revision history](../{{site.versions["stable"]}}/take-backups-with-revision-history-and-restore-from-a-point-in-time.html) may record incorrect metadata. The presence of this incorrect metadata could cause a subsequent [`RESTORE`](../{{site.versions["stable"]}}/restore.html) operation to write incorrect data:
```

Author: This isn't accurate, the first circumstance could affect either version but the second circumstance can only affect 22.2.

> - A backup of an `UPDATE`-heavy workload captures more than `kv.bulk_sst.max_allowed_overage` of revision history for a given row (64 MB by default). A subsequent `RESTORE` operation could erroneously omit or include a row, or could restore an incorrect version of it.

Reviewer: Suggested tweaks on phrasing to enable a cluster settings link, and on the default units (MB -> MiB) as specified in that doc.

```suggestion
- A backup of an [`UPDATE`](../{{site.versions["stable"]}}/update.html)-heavy workload captures more revision history for a given row than is specified by the `kv.bulk_sst.max_allowed_overage` [cluster setting](../{{site.versions["stable"]}}/cluster-settings.html) (`64 MiB` by default). A subsequent `RESTORE` operation may erroneously omit or include a row, or may restore an incorrect version of it.
```

> - In a cluster running CockroachDB v22.2, the non-public cluster setting `admission.elastic_cpu.enabled` is set to `true`. A subsequent `RESTORE` operation may erroneously omit or include any key with revision history in the backup.

Reviewer: ```suggestion
- In a cluster running CockroachDB v22.2, the non-public [cluster setting](../{{site.versions["stable"]}}/cluster-settings.html) `admission.elastic_cpu.enabled` is set to `true`. A subsequent `RESTORE` operation may erroneously omit or include any key with revision history in the backup.
```

Author: I'm not sure it's a good idea to link to the page that shows you how to set a cluster setting when we are talking about a non-public cluster setting you shouldn't be setting. WDYT?

Reviewer: That makes sense. However the way this is written leaves open the suggestion that the cluster setting can be changed to `false`. Do we explicitly recommend against this? If so, I think it would be good to say that, along with excluding the docs link.

Author: PTAL at the latest version.

> For a more detailed and technical description of the issue, refer to Github issue [#101963](https://github.com/cockroachdb/cockroach/issues/101963).

Reviewer: This seems redundant with the link in the following heading, so I'd consider cutting it.

```suggestion
For a more detailed and technical description of the issue, refer to GitHub issue [#101963](https://github.com/cockroachdb/cockroach/issues/101963).
```

This vulnerability affects all of the following CockroachDB versions: v22.2.0 to v22.2.8 and v22.1.0 to 22.1.19

## Statement

> This is resolved in CockroachDB by PR [#102179](https://github.com/cockroachdb/cockroach/pull/102179), which ensures that `RESTORE` operations can correctly restore from backups that include this incorrect metadata.

Reviewer: ```suggestion
This is resolved in CockroachDB by [#102179](https://github.com/cockroachdb/cockroach/pull/102179), which ensures that `RESTORE` operations can correctly restore from backups that include this incorrect metadata.
```

The fix has been applied to maintenance releases of CockroachDB: [v22.2.9](../releases/v22.2.html#v22-2-9).

This fix will be applied to maintenance releases of CockroachDB: v22.1.20. Until that version is available, refer to [Mitigation](#mitigation) for a suggested workaround.

This public issue is tracked by [#101963](https://github.com/cockroachdb/cockroach/issues/101963).

## Mitigation

Users of CockroachDB v22.2.0 to v22.2.8 are encouraged to upgrade to [v22.2.9](../releases/v22.2.html#v22-2-9) or a later version. After the upgrade, `RESTORE` operations will succeed even if the backup includes incorrect revision history metadata. Until such an upgrade, avoid restoring backups with revision history on the earlier version, and do not change the `admission.elastic_cpu.enabled` cluster setting from its default.

When it is available, users of CockroachDB v22.1.0 to v22.1.19 are encouraged to upgrade to v22.1.20 or later. After the upgrade, `RESTORE` operations will succeed even if the backup includes incorrect revision history metadata. Until such an upgrade, avoid restoring backups with revision history on the earlier version.

## Impact

> In specific circumstances, a `RESTORE` operation may restore incorrect data from a backup that contains incorrect metadata related to [revision history](../stable/take-backups-with-revision-history-and-restore-from-a-point-in-time.html).

Reviewer: ```suggestion
In specific circumstances, a `RESTORE` operation may restore incorrect data from a backup that contains incorrect metadata related to [revision history](../{{site.versions["stable"]}}/take-backups-with-revision-history-and-restore-from-a-point-in-time.html). This vulnerability affects all of the following CockroachDB versions: v22.2.0 to v22.2.8 and v22.1.0 to 22.1.19.
```
